DR_0008_role-based-permissions.md

```diff
@@ -0,0 +1,187 @@
+
+##### Core Mixins
+A  number of additional mixins have been added to `cadasta/core/mixins.py`. The `BaseRolePermissionMixin` checks the permissions required to access a view, sets a `_roles` attribute on the view class to hold the users current set of roles and provides a `permissions` property which composes a set of permissions based on the users currently assigned roles. The `RolePermissionRequiredMixin` extends `BaseRolePermissionsMixin` and the default django `auth.PermissionRequiredMixin`. At the moment this checks if the user has permssion to perform an action, but once `django-tutelary` has been completely removed this mixin will delegate to a custom auth backend. This is required to allow the new permissions system and `django-tutelary` to co-exist for the moment. The `RoleLoginPermissionRequiredMixin` checks if the user is authenticated and redirects accordingly. The `APIPermissionRequiredMixin` performs permission checking for `API` views.
+
+##### Organization View Mixins
+The `ProjectQuerySetMixin` has been removed and replaced with `ProjectListMixin` to provide role-based filtering of projects for `UI` and `API` project list views.  Other than this there are minimal changes to organization mixins.
+
+##### Views
+
+The existing `django-tutelary` approach to determining the required permissions for a view has been replaced with a more conventional django approach. Views can set permissions in one of three ways:
+1. Set the permission using the `permission_required` class attribute.
+2. Set `permission_required` to a dict which maps a single permission to a request method.
+3. Override `get_permission_required` if a more complex approach to determining permissions is needed, e.g. if different permissions are required based on the request method and/or organization or project visibility.
+
+##### Code examples
+
+The following code examples are indicative of the approach and subject to change:
+
+##### Groups and Permissions
+
+We define a set of platform-wide Permissions, eg.,
+
+```python
+p1 = Permission(
+		name=’View Project’, content_type=’Project’, codename=’project.view’
+	 )
+```
+
+We create the following platform-wide Groups:
+
+`PublicUser`, `ProjectUser`, `DataCollector`, `ProjectManager`, `OrgAdmin`, `SuperUser` eg:
+
+```python
+project_user = Group(name=’ProjectUser’, permissions=[p1,...,pn])
+```
+
+##### Roles
+
+We define an abstract base model Role with the following attributes:
+
+```python
+class Role (model.Model):
+	class Meta:
+	    abstract = True
+		user = ForeignKey(‘accounts.User’)
```

Comment:
As [mentioned in Cadasta/cadasta-platform#1667](https://github.com/Cadasta/cadasta-platform/pull/1667#discussion_r129706188), I think the currently proposed method of determining permissions on the Dashboard view will be necessary before we merge this into production. As currently written, I think there will be some serious impacts to performance for a user that is an `OrgAdmin` for many orgs or has `ProjectRole` permissions for many projects.